akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala

```diff
@@ -3,190 +3,263 @@
+                ).mkString(", "))
+            }
+            changedKeys.foreach(notifySubscribersFor(_, newState))
+            next(newState)
+          } else {
+            Behaviors.same
+          }
 
-    replicator ! Replicator.Subscribe(ReceptionistKey, replicatorMessageAdapter.toUntyped)
+        case NodeRemoved(address) ⇒
+          // ok to update from several nodes but more efficient to try to do it from one node
+          if (cluster.state.leader.contains(cluster.selfAddress)) {
+            nodesRemoved(Set(address))
+          } else Behaviors.same
 
-    // remove entries when members are removed
-    val clusterEventMessageAdapter: ActorRef[MemberRemoved] =
-      ctx.messageAdapter[MemberRemoved] {
-        case MemberRemoved(member, _) ⇒
+        case RemoveTick ⇒
           // ok to update from several nodes but more efficient to try to do it from one node
           if (cluster.state.leader.contains(cluster.selfAddress)) {
-            if (member.address == cluster.selfAddress) NodesRemoved.empty
-            else NodesRemoved(Set(member.address))
+            val allAddressesInState: Set[Address] = state.map.entries.flatMap {
+              case (_, values) ⇒
+                // don't care about local (empty host:port addresses)
+                values.collect { case ref if ref.path.address.hasGlobalScope ⇒ ref.path.address }
+            }(collection.breakOut)
+            val clusterAddresses = cluster.state.members.map(_.address)
+            val diff = allAddressesInState diff clusterAddresses
+            if (diff.isEmpty) Behavior.same
+            else nodesRemoved(diff)
           } else
-            NodesRemoved.empty
+            Behavior.same
       }
 
-    cluster.subscribe(clusterEventMessageAdapter.toUntyped, ClusterEvent.InitialStateAsEvents, classOf[MemberRemoved])
-
-    // also periodic cleanup in case removal from ORMultiMap is skipped due to concurrent update,
-    // which is possible for OR CRDTs
-    val removeTickMessageAdapter: ActorRef[RemoveTick.type] =
-      ctx.messageAdapter[RemoveTick.type] { _ ⇒
-        // ok to update from several nodes but more efficient to try to do it from one node
-        if (cluster.state.leader.contains(cluster.selfAddress)) {
-          val allAddressesInState: Set[Address] = state.map.entries.flatMap {
-            case (_, values) ⇒
-              // don't care about local (empty host:port addresses)
-              values.collect { case ref if ref.path.address.hasGlobalScope ⇒ ref.path.address }
-          }(collection.breakOut)
-          val clusterAddresses = cluster.state.members.map(_.address)
-          val diff = allAddressesInState diff clusterAddresses
-          if (diff.isEmpty) NodesRemoved.empty else NodesRemoved(diff)
-        } else
-          NodesRemoved.empty
+      Behaviors.immutable[Any] { (ctx, msg) ⇒
+        msg match {
+          // FIXME add union types to Scala
```

Comment:
heh ;) True but not really actionable, remove fixme as shows up in grepping for things to work on? Keep comment as explanation why we're doing this I guess

Reply:
Yeap, sorry, I was just trying to be funny.

Reply:
Union types + narrow will be awesome for public + internal protocol though.

Reply:
Indeed, would be awesome